### PR TITLE
Merge pavelvasev import related commits

### PR DESCRIPTION
**Pull request by henrikrudstrom**
Started merging pavelvasev-fork, the file Import/Include code path mainly. 
Not tested and still some commits to go. 
Comments are welcome, especially from @pavelvasev

### CODE REVIEW

**Comment by pavelvasev** [comment]
How do you do the trick of mapping these changes to new source tree layout? ;-) Manually or in some automated way?..

**Comment by henrikrudstrom** [comment]
just learned that trick: use `git cherry-pick` to pick the commits you want, if the file is renamed/split you have to move them manually to the new files (usually for qtcore.js)
```
git checkout <branch to merge to>
git checkout -b <new branch name> 
git cherry-pick <commit> 
if there is a conflict i resolve it manually here and then
git cherry-pick --continue
```

**Comment by ChALkeR** [comment]
@pavelvasev Do these fix directory imports (directory listing qmldir)?
@henrikrudstrom If they work now, could you add a testcase for that? This PR (the same branch) would be fine.

**Comment by henrikrudstrom** [comment]
Merged in my latest testing branch, 
there are still some errors (AnimatedImage, Font and one of the Async tests), but there are still quite some commits i havent merged yet.

**Comment by ChALkeR** [comment]
@henrikrudstrom I don't think that merging in all the initilization tests here is a good idea, I meant only the actual `qmldir`-based import tests.

**Comment by henrikrudstrom** [comment]
@ChALkeR wanted to make sure it doesnt break something, which it apparently does...
is there a qmldir based import test?

**Comment by pavelvasev** [comment]
It seems travis build is broken, because there are different understandings of what "import" should do.

1. In mine version, it calculates qmldir file location accourding to qml rules (including import dirs list), and tries to download it.
2. But in current main version, there are some modules already defined.

So we should skip qmldir load for some modules. 

I did it by setting `qmldirsContents` engine variable during initialization of Qmlengine object to
```
    this.qmldirsContents = { "QtQuick":{} };
``
But maybe we need more general way to consider built-in modules.. 

p.s. @ChALkeR it does only **Object Type Declaration** part of http://doc.qt.io/qt-5/qtqml-syntax-directoryimports.html#directory-listing-qmldir-files

**Comment by ChALkeR** [comment]
@pavelvasev 
> it does only **Object Type Declaration** part of http://doc.qt.io/qt-5/qtqml-syntax-directoryimports.html#directory-listing-qmldir-files

Yes, that's great!

**Comment by ChALkeR** [comment]
@henrikrudstrom 
> is there a qmldir based import test?

No, there isn't atm.

**Comment by pavelvasev** [comment]
May someone skilled perform the merge?

**Comment by ChALkeR** [comment]
@pavelvasev This is still labeled as «work in progress».

@henrikrudstrom What's the status of this? Any more work needs to be done?

**Comment by pavelvasev** [comment]
@ChALkeR @henrikrudstrom sorry, I missed the meaning of work in progress label.

As for me there are two cases:
1. If this PR is for merging things I did for importing in other qmlweb fork, that's it.
2. If this PR is for importing at all, it is not complete: it need 2.1. concerning the logic of built-in modules created with registerQmlType and 2.2. overall documentation.

My opinion is to go to case 1 (e.g. merge now) and then create different PR for case 2.

**Comment by ChALkeR** [comment]
@pavelvasev I would prefer this to be limited to import-related commits, and it looks like @henrikrudstrom also meant to do only that in this PR (judging from the PR title).

**Comment by henrikrudstrom** [comment]
@pavelvasev, thanks for picking that up. 
@ChALkeR, thats was my intention.
I think we should merge it now and ask questions later, imho the whole loading/importing part is in real need of some restructuring in the near future.
how do you want it squashed?

**Comment by ChALkeR** [comment]
@henrikrudstrom I'm removing the wip label then.
I don't want this sqashed, because it would be easier to track what changes are merged and what are not this way.

Will take a look soon.

File: src/qtcore/qml/QMLEngine.js [hunk 1]
```diff
@@ -88,11 +82,18 @@ QMLEngine = function (element, options) {
...
     }
 
+    this.extractBasePath = function( file ) {
+       var basePath = file.split(/[\/\\]/); // work both in url ("/") and windows ("\", from file://d:\test\) notation
+       basePath[basePath.length - 1] = "";
+       basePath = basePath.join("/");
+       return basePath;
+    }
     // Load file, parse and construct (.qml or .qml.js)
+    //TODO-MERGE check if this is merged correctly
```
Comment (ChALkeR) [inline src/qtcore/qml/QMLEngine.js]: This line would be removed?

Comment (ChALkeR) [inline src/qtcore/qml/QMLEngine.js]: Or should we keep it in `master` until everything gets merged? What's the intention here?

File: tests/QMLEngine/qml/somelib/SimpleButton.qml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import QtQuick 2.0
+import QtQuick.Controls 1.2
+
+Button {
+  text: "I'm simple"
+}
```
Comment (ChALkeR) [inline tests/QMLEngine/qml/somelib/SimpleButton.qml]: Newline at the end of file.

**Comment by ChALkeR** [comment]
The tests look good except for the minor style nit.
For the actual changes, I would prefer @akreuzkamp to take a quick look here.

**Comment by ChALkeR** [comment]
Waiting for @akreuzkamp review.

I really want to get this in, but I am not sure that I am qualified enough to review this, I'm not fluent with the parser and loading parts.

**Comment by ChALkeR** [comment]
Hm. I suppose we should merge this without delaying it too much. 

@pavelvasev, @henrikrudstrom, what about those two TODO comments here?

**Comment by pavelvasev** [comment]
@ChALkeR I think ) some of TODOs are obsolete, which needs investigation, and 2) others offer additional work in future.

**Comment by ChALkeR** [comment]
@pavelvasev  I meant the two merge-related ones that were introduced here.

**Comment by pavelvasev** [comment]
@ChALkeR I see only 1 TODO comment introduced here:

> //TODO-MERGE check if this is merged correctly

And I don't understand how may I help here. I already merged master somewhere in the middle and it merged ok..

**Comment by henrikrudstrom** [comment]
I also we shouldnt wait to long with merging this
LGTM

**Comment by ChALkeR** [comment]
@pavelvasev Do I get it right that we should just remove that comment?

**Comment by pavelvasev** [comment]
@chalker I think yes.

**Comment by ChALkeR** [comment]
Ok, I'm going to go ahead and merge this.

**Comment by ChALkeR** [comment]
Commits here:

93b848a addImportPath etc
6e248dc engine::setImportPathList method
4466972 bugfix for engine::addImportPath
a874cbf Allow Qt.createComponent to load from urls too
baf7a61 Bugfix in readQmlDir
d70f6ae Comment on how import statements work
8c33932 Keep info that import statement uses qualified dotted notation
349b05b Initial import implementation
96b5c2b addModulePath method and some fixes
313ac5a Import statemement test and small bugfix
ea0479a Create engine.qmldirs and engine.qmldirsContent vars earlier
bc2385e Qt.resolvedUrl function
5087bd2 Bugfix
dcb29fd convert engine calls to this calls
733a14f notes on engine::loadImports
7d338df readQmlDir - fix qmldir file url calculation
411ddd8 allow getUrlContents to optionally catch network exceptions
6eefd4e hack loadImports to skip QtQuick and QtQuick.Controls loading
13f24c9 qmldir test
d8df822 Another bugfix for resolvedUrl, and comments
3f22a38 absolute pathes in qmldir files
3264bb0 createComponent bugfix
7856f2d lookup qml files in failed dir of import
8001268 createComponent bugfix
c93baaa $resolvePath bugfix
2b42547 consider blob: in $resolvePath

**Comment by ChALkeR** [comment]
Commits here:

93b848a addImportPath etc
6e248dc engine::setImportPathList method
4466972 bugfix for engine::addImportPath
a874cbf Allow Qt.createComponent to load from urls too
baf7a61 Bugfix in readQmlDir
d70f6ae Comment on how import statements work
8c33932 Keep info that import statement uses qualified dotted notation
349b05b Initial import implementation
96b5c2b addModulePath method and some fixes
313ac5a Import statemement test and small bugfix
ea0479a Create engine.qmldirs and engine.qmldirsContent vars earlier
bc2385e Qt.resolvedUrl function
5087bd2 Bugfix
dcb29fd convert engine calls to this calls
733a14f notes on engine::loadImports
7d338df readQmlDir - fix qmldir file url calculation
411ddd8 allow getUrlContents to optionally catch network exceptions
6eefd4e hack loadImports to skip QtQuick and QtQuick.Controls loading
13f24c9 qmldir test
d8df822 Another bugfix for resolvedUrl, and comments
3f22a38 absolute pathes in qmldir files
3264bb0 createComponent bugfix
7856f2d lookup qml files in failed dir of import
8001268 createComponent bugfix
c93baaa $resolvePath bugfix
2b42547 consider blob: in $resolvePath

**Comment by ChALkeR** [comment]
I'm planning on merging these as-is, without squashing.
I will merge these with a mege commit, not with a fast-forward.
Only the merge commit will contain PR-URL and Reviewed-By headers.

**Comment by ChALkeR** [comment]
On a second though, I'm squashing some of this.

**Comment by ChALkeR** [comment]
Merged, thanks all =).

**Pull request merged**